### ISSUE TITLE
KeyCloak role mapping

### ISSUE THREAD

**Issue by LecrisUT** [zorn-v/nextcloud-social-login#209]
In case anyone is having trouble mapping roles using Keycloak, I have found the configuration needed:
Basically in the keycloak role mapper make sure to enable `Add to ID token`, e.g. using the built-in mapper:
- `Client Scopes`->`roles`->`Mappers`->`client roles`-> Enable: `Add to ID token`
And then make sure the Nextcloud configuration uses the appropriate settings:
- Groups Claim: `resource_access.${client_id}.roles`
- Group mapping: Map each role from Keycloak to corresponding Nextcloud group

I do have some question/suggetions:
- Is it possible to have a fallback configuration that any roles claimed with the same name as the Nextcloud groups are automatically mapped?
- Is it possible to set a root url for the OAuth service and have all the other urls relative to that?

**Comment by zorn-v** [comment]
> Is it possible to have a fallback configuration that any roles claimed with the same name as the Nextcloud groups are automatically mapped?

Not good idea. What if user in keycloak will have "admin" role but you don't need any admin in keycloak become admin in nextcloud ? If you need - just map it.

**Comment by LecrisUT** [comment]
> Not good idea. What if user in keycloak will have "admin" role but you don't need any admin in keycloak become admin in nextcloud ? If you need - just map it.

The roles in KeyCloak are not the same as the keycloak's admin roles etc. In the way I presented above, each role is manually created and they have to be assigned to the keycloak users/groups. So we can assume that the admin role in `resource_access.${client_id}.roles` is the intended nextcloud admin role.

If it can be toggled we can assume that the admin knows what they're doing and wishes for such an option for the sake of convenience.

**Comment by lutoma** [comment]
> Is it possible to have a fallback configuration that any roles claimed with the same name as the Nextcloud groups are automatically mapped?

+1 on that. We have a significant amount of groups mapped from Keycloak, and having to define them in two places in Nextcloud is a huge hassle.

**Comment by zorn-v** [comment]
You can use `auto_create_groups` setting

**Comment by LecrisUT** [comment]
> You can use `auto_create_groups` setting

One problem with that is that the group is only created when a user belonging to that group logs in. So we will have to have a dummy user belonging to all groups and control it to log-in in order to create the missing groups, at which point it would be less hassle to create the maps manually.

It would be more convenient to make the groups, quotas, access, hierarchy, etc. directly in Nextcloud and have the keycloak groups automatically mapped. The purpose of this issue is for simplifying the group mapping, rather than creating the missing groups. As for the security concern, at least in Keycloak it is designed safely enough that you would not assign unwanted privileges.

**Comment by zorn-v** [comment]
Got it. Update to v3.5.1 - there is automapping according to the following scheme: if group `INTERNAL_NAME-GROUP_NAME` exists it will be automapped (name the same as created by `auto_create_groups`).
For example if your provider have `keycloak` internal name, and user have `myrole` group, 
it will be auto assigned to `keycloak-myrole` nextcloud group if it exists.

**Comment by zorn-v** [comment]
By the way it work the same before with `auto_create_groups` set :smile: 
Now this option is not necessary for existing groups

**Comment by LecrisUT** [comment]
Thank you for the quick implementation. This would help with the managing.
In Nextcloud is it possible to create group aliases? I am thinking about situations like having both social-login and LDAP with the same group names. There would be duplicate `myrole` and `keycloak-myrole`.

If it's not possible to make group aliases, an alternative solution is to be able to customize `$newGroupPrefix.$`.

**Comment by zorn-v** [comment]
> In Nextcloud is it possible to create group aliases?

No. But this is the same what manual group mapping do, no ?

**Comment by LecrisUT** [comment]
Let me confirm, which implementation is right now?
|                       | implementation 1            | implementation 2         | original suggestion         |
|--------------|-------------------------|------------|-----|
| In Nextcloud | group=`keycloak-groupA` | group= `groupA` | group= `groupA` |
| In Keycloak   | role=`groupA`           | role=`keycloak-groupA`   | role= `groupA` |
| LDAP   | group=`groupA` or `keycloak-groupA`?          | group=`groupA`   | group= `groupA` |

If it's implementation 1, you can see my concern for example if we have a third service like LDAP. If it's implementation 2 than we can work with it no problem and you can disregard my previous comment.

**Comment by zorn-v** [comment]
Implementation 1
If you can manage ldap groups then what problem to set it like `keycloak-groupA` ?
Main purpose of prefix is to not allow external OAuth/OIDC service (it is not trusted source) to affect (maybe accidentally or by misconfiguration) existing nextcloud entities. Like admin user or group.

**Comment by LecrisUT** [comment]
Hmm, the same argument would apply for ldap which would create `ldap-groupA`. But I guess oauth is intended to be plubic while ldap wouldn't so it is not exactly the same. I just realized that the same configuration applies for the preconfigured ones as well, which I guess is part of your concern.

In the case of custom oauth, can you add the ability to specify the prefix including the separator? Like that if we really want to achieve the original proposition, it would be possible, including mapping to admin.

**Comment by zorn-v** [comment]
Second purpose to not mix data from several providers if so.

> In the case of custom oauth, can you add the ability to specify the prefix including the separator?

Make sense. And if it empty just use current behavior.

**Comment by zorn-v** [comment]
> In the case of custom oauth

Custom also has no trust

**Comment by TBG-FR** [comment]
So there is no way to avoid manually mapping each groups on both sides ? I thought with the `Groups Claim` settings I'd just had to create and assign roles in Keycloak for them to work in Nextcloud...

**Comment by zorn-v** [comment]
Why you need to map on both sides ?
Map is needed to map ONE side to ANOTHER. If you need to map both sides, may be you do not need mapping ?

**Comment by TBG-FR** [comment]
> Why you need to map on both sides ? Map is needed to map ONE side to ANOTHER. If you need to map both sides, may be you do not need mapping ?

I would like my Keycloak groups reflected inside Nextcloud, so each user groupe/role can share a specific common space

**Comment by zorn-v** [comment]
I would like peace in the world, and so what ?
Why keykloak should change something in nextcloud ?

**Comment by TBG-FR** [comment]
> I would like peace in the world, and so what ? Why keykloak should change something in nextcloud ?
> 
> PS. Sorry. Can you give some tech ask how to consider it ? PPS. And doesn not broke what works

No need to be sarcastic or condescending....

I just don't understand why we couldn't have dynamic mapping, but my need and/or understanding of the subject is probably different from yours

- Can you just confirm that today we're forced to map (1 to 1, manually) keycloak groups and nextcloud groups
- If yes, then why couldn't we map them dynamically (1 to 1, but without manually doing it) ? Keycloak is, for a lot of us, the source of truth, so we'd like to rely on that 😅 
- Maybe I'm wrong, but I feel like it's not "keycloak should change something in Nextcloud", it's "automate what exists today"

I didn't understand your PS, but yes I would like to contribute and add this feature, but I'm not sure to be able to find the time to, so I want to get your approval before doing anything

**Comment by zorn-v** [comment]
> If yes, then why couldn't we map them dynamically

You can. Just enable auto_create_groups option

**Comment by TBG-FR** [comment]
> > If yes, then why couldn't we map them dynamically
> 
> You can. Just enable auto_create_groups option

Thank you for your answer ! However, with `auto_create_groups` enabled, my users are not given any group...

## Information
- Nextcloud version : `Nextcloud Hub 3 (25.0.4) stable`
- Social login version : `5.4.3`
- Custom OpenID connection with correct information
- Auto create groups if they don't exist : `true` ✅ 
- Groups claim : `resource_access.drive-nextcloud.roles`

## Example token content
```
{
  [...]
  "azp": "drive-nextcloud",
  "realm_access": {
    "roles": [
      "offline_access",
      "realm_role_03",
      "uma_authorization",
      "default-roles-realmname"
    ]
  },
  "resource_access": {
    "drive-nextcloud": {
      "roles": [
        "documents_level_aaa",
        "documents_level_bbb"
      ]
    },
    "account": {
      "roles": [
        "manage-account",
        "manage-account-links",
        "view-profile"
      ]
    }
  },
  "scope": "email profile",
  [...]
  "name": "User NAME",
  "preferred_username": "user.name",
  "given_name": "User",
  "family_name": "NAME",
  "email": "blabla@example.com"
}
```

## Behaviour
- Users are able to connect using Keycloak
- **They're not put in any group**
- **No new group is created**

Anything I could have missed ?

**Comment by zorn-v** [comment]
> Users are able to connect using Keycloak

Connect (you mean that one in personal settings, right ?) is do only one thing - connect social login to nextcloud account, that's all. No other things like `update_profile_on_login`, creating/mapping groups etc.
I don't think that it is good idea to change that behavior

**Comment by TBG-FR** [comment]
Let me rephrase it : "Users are able to login, using the `Login with Keycloak SSO` button" but they're not assigned any group, and no group is created

What is `auto_create_groups` supposed to do ? I thought it would create groups on login, based on `groups claim` value...

I don't want to change behaviours, I'm trying to understand how it works 😅

**Comment by zorn-v** [comment]
You should also enable "Update user profile every login" for sync groups for already created users.

**Comment by zorn-v** [comment]
Yes, it should.
Try to add something like
```php
var_dump($profile->data); die();
```
in `lib/Service/ProviderService.php:507` line and try to login to check is groups correctly filled from response.

**Comment by TBG-FR** [comment]
Alright, it was a Keycloak issue 🤦‍♂️ Looks like it was not a configuration issue, but a bug

## Related links
https://stackoverflow.com/a/62925130/8179249
https://keycloak.discourse.group/t/resource-access-claim-missing-from-userinfo-until-i-change-the-name/1238
https://issues.redhat.com/browse/KEYCLOAK-9874
https://github.com/keycloak/keycloak/pull/7408

## How to fix it
Updating Keycloak to a recent version should do it, but there is also a simple workaround 
- Login to Keycloak with an admin account
- Go to `Client scopes` > `roles` -> `mappers` > `client roles`
- Copy and cut the field `Token Claim Name`, then save
- Paste the copied data (exactly the same) in that field again, then save
- Now it's working like a charm 🥳 

Thank you for your time 🙏  sorry for all the noise 😅

**Comment by ghorchem** [comment]
Is there a way to configure Group Mapping using the command line in Ubuntu? I’ve tried using the gui and when I click on save it deletes the custom provider. Igor that part configured using the command line.
